components/eam/src/physics/crm/samxx/setperturb.cpp

```diff
@@ -1,47 +1,39 @@
-
-  real2d t02 = real2d("t02",perturb_num_layers,ncrms);
-  yakl::memset(t02,0.);
-
+  real factor_xy = 1. / (nx*ny);
+  auto t_host     = t    .createHostCopy();
+  auto t0_host    = t0   .createHostCopy();
+  auto gcolp_host = gcolp.createHostCopy();
   // Apply random liquid static energy (LSE) perturbations
-  // do k = 1,perturb_num_layers
-  //   do j = 1,ny
-  //     do i = 1,nx
-  //       do icrm = 1 , ncrms
-  auto ncrms = ::ncrms;
-  auto t     = ::t;
-  auto t0    = ::t0;
-  auto gcolp = ::gcolp;
-  parallel_for( SimpleBounds<4>(perturb_num_layers,ny,nx,ncrms) , YAKL_LAMBDA (int k, int j, int i, int icrm) {
-    // set perturb_k_scaling so that perturbation magnitude decreases with altitude
-    real perturb_k_scaling = ((real)perturb_num_layers-k) / (real)perturb_num_layers;
-    // Get the random number
-    yakl::Random rand( (size_t) ( gcolp(icrm)*perturb_num_layers*ny*nx*ncrms + 
-                                  k*ny*nx*ncrms + j*nx*ncrms + i*ncrms + icrm ) );
-    real rand_perturb = rand.genFP<real>( -1. , 1. );
-    // apply perturbation 
-    t(k,j+offy_s,i+offx_s,icrm) = t(k,j+offy_s,i+offx_s,icrm) + rand_perturb * perturb_t_magnitude * perturb_k_scaling;
-    // Calculate new average LSE for energy conservation scaling below
-    yakl::atomicAdd( t02(k,icrm) , t(k,j+offy_s,i+offx_s,icrm)/( (real) nx * (real) ny ) );
-  });
-
-  // enforce energy conservation
-  // do k = 1,perturb_num_layers
-  //   do j = 1,ny
-  //     do i = 1,nx
-  //       do icrm = 1 , ncrms
-  parallel_for( SimpleBounds<4>(perturb_num_layers,ny,nx,ncrms) , YAKL_LAMBDA (int k, int j, int i, int icrm) {
-    t(k,j+offy_s,i+offx_s,icrm) = t(k,j+offy_s,i+offx_s,icrm) * t0(k,icrm) / t02(k,icrm);
-  });
-}
-
+  for (int icrm = 0; icrm < ncrms; icrm++) {
+    srand(gcolp_host(icrm)*1000);
+    for (int k = 0; k < perturb_num_layers; k++) {
+      // set perturb_k_scaling so that perturbation magnitude decreases with altitude
+      real perturb_k_scaling = ((real)perturb_num_layers-k) / (real)perturb_num_layers;
+      real t02 = 0;
+      for (int j = 0; j < ny; j++) {
+        for (int i = 0; i < nx; i++) {
+          // Generate a uniform random number in interval (0,1)
+          real rand_perturb = static_cast<double>( rand() ) / static_cast<double>( RAND_MAX );
+          // onvert perturbation range from (0,1) to (-1,1)
+          rand_perturb = 1.-2.*rand_perturb;
+          // apply perturbation 
+          t_host(k,j+offy_s,i+offx_s,icrm) = t_host(k,j+offy_s,i+offx_s,icrm) + rand_perturb * perturb_t_magnitude * perturb_k_scaling;
+          // Calculate new average LSE for energy conservation scaling below
+          t02 += t_host(k,j+offy_s,i+offx_s,icrm)*factor_xy;
+        }
+      }
+      for (int j = 0; j < ny; j++) {
+        for (int i = 0; i < nx; i++) {
+          t_host(k,j+offy_s,i+offx_s,icrm) = t_host(k,j+offy_s,i+offx_s,icrm) * t0_host(k,icrm) / t02;
+        }
+      }
+    }
+  }
+  t_host.deep_copy_to(t);
+}
```

Comment:
Missing end of file line.

Reply:
Did you mean for this change to the initial perturbations to be part of the PR?

Reply:
Done. Yes, this fix was important for Vt. I'd like to keep them together for this instance rather than separating this out into a separate PR.